[mlir][linalg] Emit a warning when tile_using_forall generates non thread-safe code

pabloantoniom opened the pull request:
**Description**

The documentation of `transform.structured.tile_using_forall` says:

_"It is the user’s responsibility to ensure that num_threads/tile_sizes is a valid tiling specification (i.e. that only tiles parallel dimensions, e.g. in the Linalg case)."_

In other words, tiling a non-parallel dimension would generate code with data races which is not safe to parallelize. For example, consider this example (included in the tests in this PR):

```
func.func @tile_thread_safety2(%arg0: tensor<100x300x8xf32>, %arg1: tensor<300x8xf32>) -> tensor<300x8xf32> {
  %0 = scf.forall (%arg2) in (8) shared_outs(%arg3 = %arg1) -> (tensor<300x8xf32>) {
    %1 = affine.min #map(%arg2)
    %2 = affine.max #map1(%1)
    %3 = affine.apply #map2(%arg2)
    %extracted_slice = tensor.extract_slice %arg0[%3, 0, 0] [%2, 300, 8] [1, 1, 1] : tensor<100x300x8xf32> to tensor<?x300x8xf32>
    %4 = linalg.generic {indexing_maps = [#map3, #map4], iterator_types = ["reduction", "parallel", "parallel"]} ins(%extracted_slice : tensor<?x300x8xf32>) outs(%arg3 : tensor<300x8xf32>) {
    ^bb0(%in: f32, %out: f32):
      %5 = arith.addf %in, %out : f32
      linalg.yield %5 : f32
    } -> tensor<300x8xf32>
    scf.forall.in_parallel {
      tensor.parallel_insert_slice %4 into %arg3[0, 0] [300, 8] [1, 1] : tensor<300x8xf32> into tensor<300x8xf32>
    }
  }
  return %0 : tensor<300x8xf32>
}
```

We can easily see that this is not safe to parallelize because all threads would be writing to the same position in `%arg3` (in the `scf.forall.in_parallel`.

This PR detects wether it's safe to `tile_using_forall` and emits a warning in the case it is not.

**Brief explanation**
It first generates a vector of affine expressions representing the tile values and stores it in `dimExprs`. These affine expressions are compared with the affine expressions coming from the results of the affine map of each output in the linalg op. So going back to the previous example, the original transform is:

```
#map = affine_map<(d0, d1, d2) -> (d0, d1, d2)>
#map1 = affine_map<(d0, d1, d2) -> (d1, d2)>

func.func @tile_thread_safety2(%arg0: tensor<100x300x8xf32>, %arg1: tensor<300x8xf32>) -> tensor<300x8xf32> {
  // expected-warning@+1 {{tiling is not thread safe at axis #0}}
  %0 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["reduction", "parallel", "parallel"]} ins(%arg0 : tensor<100x300x8xf32>) outs(%arg1 : tensor<300x8xf32>) {
  ^bb0(%in: f32, %out: f32):
    %1 = arith.addf %in, %out : f32
    linalg.yield %1 : f32
  } -> tensor<300x8xf32>
  return %0 : tensor<300x8xf32>
}

module attributes {transform.with_named_sequence} {
  transform.named_sequence @__transform_main(%arg0: !transform.any_op {transform.readonly}) {
    %0 = transform.structured.match ops{["linalg.generic"]} in %arg0 : (!transform.any_op) -> !transform.any_op
    %forall, %tiled_generic = transform.structured.tile_using_forall %0 num_threads [8]
          : (!transform.any_op) -> (!transform.any_op, !transform.any_op)
    transform.yield
  }
}
```

The `num_threads` attribute would be represented as `(d0)`. Because the linalg op has only one output (`arg1`) it would only check against the results of `#map1`, which are `(d1, d2)`. The idea is to check that all affine expressions in `dimExprs` are present in the output affine map. In this example, `d0` is not in `(d1, d2)`, so tiling that axis is considered not thread safe.

pabloantoniom committed:
[mlir][Linalg][Transform] Emit a warning when tile_using_forall generates non thread-safe code

This warning aims to complement the comment in the documentation that
says:

"It is the user’s responsibility to ensure that num_threads/tile_sizes
is a valid tiling specification (i.e. that only tiles parallel
dimensions, e.g. in the Linalg case)."

because:

1. Not all users of tile_using_forall know that tiling the wrong
dimension/s (e.g., a non-parallel dimension) will generate non
thread-safe code, so this warning will inform the user about it.

2. Users of tile_using_forall may know this limitation, but they may
not realize that they are tiling a non-parallel dimension, so the
warning may help in the debugging process.

pabloantoniom commented:
Ping

pabloantoniom commented:
Ping

joker-eph commented:
Thanks for spending effort on describing what you're doing in the description, however:

> This PR detects whether it's safe to tile_using_forall and emits a warning in the case it is not.

I don't find in the description a motivation for this change right now. Why is a warning a good thing here?

pabloantoniom commented:
> Thanks for spending effort on describing what you're doing in the description, however:
> 
> > This PR detects whether it's safe to tile_using_forall and emits a warning in the case it is not.
> 
> I don't find in the description a motivation for this change right now. Why is a warning a good thing here?

The motivation behind this change is to complement the comment in the documentation that says:

"_It is the user’s responsibility to ensure that num_threads/tile_sizes is a valid tiling specification (i.e. that only tiles parallel dimensions, e.g. in the Linalg case)._"

because:

1. Not all users of `tile_using_forall` know that tiling the wrong dimension/s (e.g., a non-parallel dimension) will generate non thread-safe code (like me a few months ago), so a warning will inform the user about it.
2. Users of `tile_using_forall` may know this limitation, but they may not realise that they are tiling a non-parallel dimension (like me a few weeks ago), so the warning may help those users in the debugging process.

ftynse commented:
This justification should be included in the commit message: https://mlir.llvm.org/getting_started/Contributing/#commit-messages.

pabloantoniom commented:
> This justification should be included in the commit message: https://mlir.llvm.org/getting_started/Contributing/#commit-messages.

My bad. I have added it now, thanks.

pabloantoniom commented:
Another option I also considered is to emit an error instead of a warning, and then add an option to `tile_using_forall` to disable this check (e.g., `check_unsafe_tiling=False`). This way we ensure that the user will not generate non thread-safe code unless the option is enabled explicitly.

pabloantoniom commented:
Ping

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -304,6 +304,50 @@ static void calculateTileOffsetsAndSizes(
+  dimExprs.reserve(numDims);
+  for (unsigned i = 0; i < tilingValues.size(); i++) {
+    if (auto attr = llvm::dyn_cast_if_present<Attribute>(tilingValues[i])) {
+      if (cast<IntegerAttr>(attr).getValue().getSExtValue() > 1)
```
This doesn't look correct given that `tilingValues` is either tile sizes or num threads (that is, number of tiles). When tile size is 1, we will have as many tiles (and, therefore, hypothetical threads in the forall loop) as elements along the axis, so it's clearly not safe to ignore here.

pabloantoniom replied:
It is correct if `num_threads` is used, since tiling with 1 thread means that only 1 thread runs, so there is no race condition. But it is incorrect in the case `tile_sizes` are used, good catch. Fixed!

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -304,6 +304,50 @@ static void calculateTileOffsetsAndSizes(
+    }
+  }
+
+  for (uint32_t resNum = 0; resNum < op->getNumResults(); resNum++) {
```
Nit: why is this specifically `uint32_t`?  `getNumResults()` returns `unsigned`.

pabloantoniom replied:
My bad, fixed.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -304,6 +304,50 @@ static void calculateTileOffsetsAndSizes(
+
+  SmallVector<AffineExpr> dimExprs;
+  dimExprs.reserve(numDims);
+  for (unsigned i = 0; i < tilingValues.size(); i++) {
```
Nit: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

pabloantoniom replied:
Fixed

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -304,6 +304,50 @@ static void calculateTileOffsetsAndSizes(
+    }
+  }
+
+  for (uint32_t resNum = 0; resNum < op->getNumResults(); resNum++) {
+    AffineMap map =
+        linalgOp.getIndexingMapMatchingResult(op->getResult(resNum));
+
+    for (AffineExpr r : dimExprs) {
+      unsigned int axis = cast<AffineDimExpr>(r).getPosition();
+      if (!llvm::is_contained(map.getResults(), r))
+        safeToTile[axis] = false;
+    }
+  }
```
I don't quite understand this. Can't we just say we shouldn't tile to forall an iterator that is marked as "reduction"? What are the semantics of a `linalg.generic` that writes to the same result multiple times without reducing, which is what would happen when an iteration space axis is not present in the result indexing map? I suspect that the behavior is undefined given the intrinsically parallel semantics of linalg and therefore having a race is just fine. But can we even construct such an operation and have it pass the verifier?

pabloantoniom replied:
I agree, we can simply check the iterator type. I have simplified my implementation to do this. 

I imagine that, just like you pointed out, the semantics of a `linalg.generic` that performs a reduction but uses the parallel iterator type are undefined. My original implementation would still catch those cases, but yes, I guess in those cases having a race is fine. I wonder, however, if my previous implementation could be reused to verify if the iterator types are correct (e.g., if the code performs a reduction but the corresponding iterator type is "parallel"). I don't know how useful/relevant that would be.

ftynse replied:
If we can indeed construct such an operation (I haven't tried), it would be useful to have a verifier.

ftynse reviewed (changes requested):
Please add a test exercising this for `tile_sizes`, not only `num_threads`. I suspect the logic is broken.

pabloantoniom committed:
[mlir][Linalg][Transform] Small nits, bugfix and tests.

Fix bug when tile_size=1 was specified. Also,
add test case using tile_size and another use
case to tile linalg.matmul to show that this
also works for non linalg.generic ops

pabloantoniom committed:
[mlir][Linalg][Transform] Simplify implementation

Rather than comparing the outputs affine maps against the maps infered
from the tile sizes, we can simply check that tiled dimensions do not
contain the "reduction" iterator type. If they do, then we are certain
they are not safe to tile.

pabloantoniom commented:
> Please add a test exercising this for `tile_sizes`, not only `num_threads`. I suspect the logic is broken.

Thanks for taking your time to review the PR. I have added a test for tile_sizes (indeed there was an issue when tile_sizes contained a 1, like you suggested in a comment below). I have also added a test case showcasing that this also works for other linalg ops, not only linalg.generic.

pabloantoniom commented on mlir/test/Dialect/Linalg/tile-to-forall.mlir:
```diff
@@ -586,3 +586,103 @@ module attributes {transform.with_named_sequence} {
+#map1 = affine_map<(d0, d1) -> (d0)>
+
+func.func @tile_thread_safety1(%arg0: tensor<100x300xf32>, %arg1: tensor<100xf32>) -> tensor<100xf32> {
+  // expected-warning@+1 {{tiling is not thread safe at axis #1}}
```
Fixed, thanks

pabloantoniom commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -304,6 +304,50 @@ static void calculateTileOffsetsAndSizes(
+  SmallVector<bool> safeToTile(tilingValues.size(), true);
+  LinalgOp linalgOp = dyn_cast<LinalgOp>(op.getOperation());
+  if (!linalgOp)
+    return safeToTile;
```
Agreed, fixed

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -304,6 +304,44 @@ static void calculateTileOffsetsAndSizes(
 }
 
+/// Returns a vector of bools representing if, for the given axis, `op` can be
+/// tiled by without incurring in a race condition and thus it is thread-safe to
```
Nit: tiled by what?

pabloantoniom replied:
Fixed

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -304,6 +304,44 @@ static void calculateTileOffsetsAndSizes(
+
+  auto iterators = linalgOp.getIteratorTypesArray();
+  for (AffineExpr r : dimExprs) {
+    unsigned int axis = cast<AffineDimExpr>(r).getPosition();
```
We no longer need affine expressions here. The code above can just push `i` in the vector, and this code can just iterate over those. Furthermore, it may not be necessary to even have the second loop and a vector to communicate between them.

pabloantoniom replied:
Agreed. I have removed the affine expressions and also merged both loops.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Tiling.cpp:
```diff
@@ -304,6 +304,44 @@ static void calculateTileOffsetsAndSizes(
+                   ArrayRef<OpFoldResult> numThreads,
+                   std::optional<ArrayRef<OpFoldResult>> nominalTileSizes,
+                   int numDims) {
+  ArrayRef<OpFoldResult> tilingValues =
+      nominalTileSizes.has_value() ? *nominalTileSizes : numThreads;
+  int minTile = nominalTileSizes.has_value() ? 0 : 1;
```
I really dislike how this mixes tile sizes and num threads in one variable. These are opposite things. Could we rather not mix them and just have two separate checks below?

pabloantoniom replied:
I feel the same. However, having duplicated code (one check for tile sizes and numThreads) does not look good to me either. However, now I have noticed that when tile sizes are specified, numThreads are automatically derived from tile sizes (in `tileToForallOpUsingTileSizes`). Therefore, we now that numThreads will always contain the values that we need to check, so I have simplified the implementation to check only numThreads.

ftynse commented on mlir/test/Dialect/Linalg/tile-to-forall.mlir:
```diff
@@ -586,3 +586,144 @@ module attributes {transform.with_named_sequence} {
+          : (!transform.any_op) -> (!transform.any_op, !transform.any_op)
+    transform.yield
+  }
+}
```
Nit: please keep the trailing newline.

pabloantoniom replied:
Fixed

ftynse reviewed (approved):
Please address the remaining comments.

pabloantoniom committed:
[mlir][Linalg][Transform] Simplify even more the implementation

Simplify the implementation by removing the end loop and by only
checking numThreads (before it was also checking nominalTileSizes).
This is possible because in the event that num_threads is not specified,
it is automatically derived from tile_sizes, so we are always sure that
numThreads will contain the values we need.

Also changed a bit the tests to show that it the implementation
properly handles the case where there is a zero.

pabloantoniom committed:
clang-format

pabloantoniom commented:
> Please address the remaining comments.

Thanks for taking the time to review the changes. I have significantly simplified the implementation, so it would be good if you could have a last look before merging. If so, please feel free to merge, otherwise I'll do it later.
